modules/s3-articfact/main.tf

```diff
@@ -0,0 +1,380 @@
+  }
+}
+
+# -------------------------------------------
+# SET PROVIDER ALIAS FOR REPLICATION
+# -------------------------------------------
+
+provider "aws" {
+  region = var.priamry_region
+}
+
+provider "aws" {
+  alias  = "replica"
+  region = var.replica_region
+}
+
+
+# ------------------------------------------------------------
+
+# THE FOLLOWING SECTION IS USED TO CREATE THE PRIMARY S3 BUCKET
+
+# AND ASSOCIATED RESOURCES (LIFECYCLE MANAGEMENT, VERSIONING).
+
+# ------------------------------------------------------------
+
+
+# -------------------------------------------
+# CREATE PRIMARY S3 BUCKET
+# -------------------------------------------
+
+resource "aws_s3_bucket" "primary" {
+  bucket = var.primary_bucket_name
+}
+
+# -------------------------------------------
+# CREATE PRIMARY S3 BUCKET POLICY
+# -------------------------------------------
+
+resource "aws_s3_bucket_acl" "primary" {
+  bucket = aws_s3_bucket.primary.id
+  acl    = var.pimary_bucket_acl
+}
+
+# -------------------------------------------
+# ENABLE PRIMARY OBJECT VERSIONING
+# -------------------------------------------
+
+resource "aws_s3_bucket_versioning" "primary" {
+  bucket = aws_s3_bucket.primary.id
+  versioning_configuration {
+    status = var.bucket_versioning_configuration_status
+  }
+}
+
+# -------------------------------------------
+# CONFIGURE PARTIAL LIFECYCLE MANAGEMENT
+# -------------------------------------------
+
+resource "aws_s3_bucket_lifecycle_configuration" "pirmary" {
+  count = !var.enable_storage_lifecycles ? 1 : 0
+
+  bucket = aws_s3_bucket.primary.id
+
+  rule {
+    id = var.lifecycle_versioning_id
+
+    status = var.bucket_versioning_configuration_status
+
+    noncurrent_version_expiration {
+      noncurrent_days           = 30
+      newer_noncurrent_versions = 1
+    }
+  }
+}
+
+# -------------------------------------------
+# CONFIGURE FULL LIFECYCLE MANAGEMENT
+# -------------------------------------------
+
+resource "aws_s3_bucket_lifecycle_configuration" "primary" {
+  count = var.full_lifecycle_management ? 1 : 0
+
+  bucket = aws_s3_bucket.primary.id
+
+  rule {
+    id = var.lifecycle_transitioin_id
+
+    status = var.bucket_versioning_configuration_status
+
+    transition {
+      days          = 30
+      storage_class = var.transition_30_storage_class
+    }
+
+    transition {
+      days          = 90
+      storage_class = var.transition_90_storage_class
+    }
+
+  }
+
+  rule {
+    id = var.lifecycle_versioning_id
+
+    status = var.bucket_versioning_configuration_status
+
+    noncurrent_version_expiration {
+      noncurrent_days           = 30
+      newer_noncurrent_versions = 1
+    }
+  }
+}
+
+# -------------------------------------------
+# CONFIGURE PRIMARY S3 BUCKET PUBLIC ACCESS
+# -------------------------------------------
+resource "aws_s3_bucket_public_access_block" "primary" {
+  bucket = aws_s3_bucket.primary.id
+
+  block_public_acls       = true
+  block_public_policy     = true
```

Comment:
block_public_policy will only be false if we `enable_public_access`. The use case for this is clark-prod-file-uploads. We have a public access policy for that bucket.